src/runtime/relax_vm/cuda/cuda_graph_builtin.cc

```diff
@@ -0,0 +1,156 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file src/runtime/relax_vm/cuda_graph_builtin.cc
+ * \brief The CUDA graph related builtin functions for Relax virtual machine.
+ */
+
+#include <tvm/runtime/container/adt.h>
+#include <tvm/runtime/data_type.h>
+#include <tvm/runtime/device_api.h>
+#include <tvm/runtime/logging.h>
+#include <tvm/runtime/memory.h>
+#include <tvm/runtime/ndarray.h>
+#include <tvm/runtime/packed_func.h>
+#include <tvm/runtime/registry.h>
+#include <tvm/runtime/relax_vm/bytecode.h>
+#include <tvm/runtime/relax_vm/memory_manager.h>
+#include <tvm/runtime/relax_vm/vm.h>
+
+#include "../../cuda/cuda_common.h"
+namespace tvm {
+namespace runtime {
+namespace relax_vm {
+
+using tvm::runtime::NDArray;
+
+class CUDAGraphNode : public Object {
+ public:
+  cudaGraph_t handle_;
+  TVM_DECLARE_FINAL_OBJECT_INFO(CUDAGraphNode, Object);
+};
+
+class CUDAGraph : public ObjectRef {
+ public:
+  explicit CUDAGraph(cudaGraph_t handle) {
+    auto n = make_object<CUDAGraphNode>();
+    n->handle_ = handle;
+    data_ = std::move(n);
+  }
+  TVM_DEFINE_OBJECT_REF_METHODS(CUDAGraph, ObjectRef, CUDAGraphNode);
+};
+
+class CUDAGraphCache : public Object {
+ public:
+  struct Entry {
+    std::vector<ObjectRef> states;
+    CUDAGraph graph;
+  };
+  std::unordered_map<ObjectRef, Entry, ObjectPtrHash, ObjectPtrEqual> entries;
+
+  static CUDAGraphCache* Get() { return dmlc::ThreadLocalStore<CUDAGraphCache>::Get(); }
+};
+
+struct CaptureContext {
+  cudaGraph_t graph;
+  cudaStream_t stream;
+  // TODO: refactor CUDA graph rewrite pass so that warm up run is not needed
+  bool is_warm_up;
+};
+
+typedef dmlc::ThreadLocalStore<CaptureContext*> CaptureContextStore;
+
+TVM_REGISTER_GLOBAL("vm.builtin.get_captured_cuda_graph")
+    .set_body([](TVMArgs args, TVMRetValue* rv) {
+      ICHECK_EQ(args.size(), 2);
+      VirtualMachine* vm = VirtualMachine::GetContextPtr(args[0]);
+      ObjectRef func = args[1];
+
+      CUDAGraphCache* cache = CUDAGraphCache::Get();
+      if (auto it = cache->entries.find(func); it == cache->entries.end()) {
+        CaptureContext ctx;
+        CUDA_CALL(cudaStreamCreate(&ctx.stream));
+        // ctx.stream = static_cast<cudaStream_t>(device_api->CreateStream(device));
+        ctx.is_warm_up = true;
+        *CaptureContextStore::Get() = &ctx;
+
+        // PackedFunc func{nullptr};
+        // func = vm->GetFunction(func_name, GetObjectPtr<Object>(vm));
+        // ICHECK(func != nullptr);
+
+        // warm up
+        TVMArgs func_args(nullptr, nullptr, 0);
```

Comment:
Would be nice to separate the allocation and capture function, so we don't need the ctx. 

The allocation and capture function separation can also be helpful for generalization to other APIs. As we no longer need begin_capture and end_capture anymore